jupyter_client/tests/test_kernelmanager.py

```diff
@@ -7,117 +7,174 @@
-                'display_name': "Signal Test Kernel",
-                'env': {'TEST_VARS': '${TEST_VARS}:test_var_2'},
-            }))
-
-    def _get_tcp_km(self):
-        c = Config()
-        km = KernelManager(config=c)
-        return km
-
-    def _get_ipc_km(self):
-        c = Config()
-        c.KernelManager.transport = 'ipc'
+@pytest.fixture(autouse=True)
+def env():
+    env_patch = test_env()
+    env_patch.start()
+    yield
+    env_patch.stop()
+
+
+@pytest.fixture(params=['tcp', 'ipc'])
+def transport(request):
+    if sys.platform == 'win32' and request.param == 'ipc':  #
+        pytest.skip("Transport 'ipc' not supported on Windows.")
+    return request.param
+
+
+@pytest.fixture
+def config(transport):
+    c = Config()
+    c.KernelManager.transport = transport
+    if transport == 'ipc':
         c.KernelManager.ip = 'test'
-        km = KernelManager(config=c)
-        return km
+    return c
+
+
+@pytest.fixture
+def install_kernel():
+    kernel_dir = pjoin(paths.jupyter_data_dir(), 'kernels', 'signaltest')
+    os.makedirs(kernel_dir)
+    with open(pjoin(kernel_dir, 'kernel.json'), 'w') as f:
+        f.write(json.dumps({
+            'argv': [sys.executable,
+                     '-m', 'jupyter_client.tests.signalkernel',
+                     '-f', '{connection_file}'],
+            'display_name': "Signal Test Kernel",
+            'env': {'TEST_VARS': '${TEST_VARS}:test_var_2'},
+        }))
+
+
+@pytest.fixture
+def start_kernel():
+    km, kc = start_new_kernel(kernel_name='signaltest')
+    yield km, kc
+    kc.stop_channels()
+    km.shutdown_kernel()
+    assert km.context.closed
+
+
+@pytest.fixture
+def start_kernel_w_env():
+    kernel_cmd = [sys.executable,
+                  '-m', 'jupyter_client.tests.signalkernel',
+                  '-f', '{connection_file}']
+    extra_env = {'TEST_VARS': '${TEST_VARS}:test_var_2'}
+
+    km = KernelManager(kernel_name='signaltest')
+    km.kernel_cmd = kernel_cmd
+    km.extra_env = extra_env
+    km.start_kernel()
+    kc = km.client()
+    kc.start_channels()
+
+    kc.wait_for_ready(timeout=60)
+
+    yield km, kc
+    kc.stop_channels()
+    km.shutdown_kernel()
+
+
+@pytest.fixture
+def km(config):
+    km = KernelManager(config=config)
+    return km
+
+
+@pytest.fixture
+def zmq_context():
+    import zmq
+    ctx = zmq.Context()
+    yield ctx
+    ctx.term()
+
+
+@pytest.fixture(params=[AsyncKernelManager, AsyncKernelManagerSubclass, AsyncKernelManagerWithCleanup])
+def async_km(request, config):
+    km = request.param(config=config)
+    return km
+
+
+@pytest.fixture
+@async_generator  # This is only necessary while Python 3.5 is support afterwhich both it and yield_() can be removed
+async def start_async_kernel():
+    km, kc = await start_new_async_kernel(kernel_name='signaltest')
+    await yield_((km, kc))
+    kc.stop_channels()
+    await km.shutdown_kernel()
+    assert km.context.closed
+
+
+class TestKernelManager:
```

Comment:
I don't think the class is necessary btw. You can just define test functions directly.

Reply:
True.  I was just keeping the classes around for organization purposes, although I did find it convenient to specify `@pytest.mark.asyncio` on the `TestAsyncKernelManager` declaration for its tests.

Do you want these removed and things flattened out?

Reply:
Nah, I agree it's helping organize here. Was just commenting in case it wasn't known. Let's merge your work :)